### ISSUE TITLE
Docker-compose

### ISSUE THREAD

**Issue by olalonde** [minio/minio#2796]
Not really an issue but thought I'd share my docker-compose setup here in case it's useful to others.

``` yaml
minio:
  image: minio/minio
  ports:
    - "9000:9000"
  volumes:
    - ./test/.minio/data:/export
    - ./test/.minio/config:/root/.minio
  environment:
    - "MINIO_ACCESS_KEY=AKIAIOSFODNN7EXAMPLE"
    - "MINIO_SECRET_KEY=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY"

createbucket:
  image: garland/aws-cli-docker
  links:
    - minio
  command: >
    /bin/sh -c "
      while ! wget -q -s http://minio:9000/minio/login; do sleep 1; done;
      aws configure set default.s3.signature_version s3v4;
      aws configure set default.region us-east-1;
      aws s3api create-bucket \
        --bucket somebucketname \
        --endpoint-url=http://minio:9000;
      exit 0;
    "
  environment:
    - "AWS_ACCESS_KEY_ID=AKIAIOSFODNN7EXAMPLE"
    - "AWS_SECRET_ACCESS_KEY=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY"
```

```
# creates your bucket
docker-compose up createbucket
# starts minio
docker-compose up -d minio
```

**Comment by KanybekMomukeyev** [comment]
@olalonde thank you very much, saved my day, in documentaion bit more complicated!

**Comment by KanybekMomukeyev** [comment]
My version is like this:

```
s3minio:
    image: minio/minio
    ports:
      - "9000:9000"
    volumes:
      - "./somefolder:/export"
    environment:
      - "MINIO_ACCESS_KEY=AKIAIOSFODNN7EXAMPLE"
      - "MINIO_SECRET_KEY=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY"
    command: server /export
```

**Comment by olalonde** [comment]
You're welcome! A recent commit removed the default command so it has to be specified in docker-compose now... Opened a PR https://github.com/minio/minio/pull/3030

**Comment by tzookb** [comment]
is there a less complex way to just create bucket in the minio container?

something like
  
    buckets:
       - bucketOne
       - bucketTwo

?

Thanks!

**Comment by harshavardhana** [comment]
> is there a less complex way to just create bucket in the minio container?
> 
> something like
> 
> buckets:
>    - bucketOne
>    - bucketTwo
> ?

You can use minio browser, you can use [mc](https://github.com/minio/mc) - compose doesn't speak S3 API so there would be no way to create buckets in that manner - Also for new issues please open a new one @tzookb. 

Better yet we hang out at https://slack.minio.io feel free to join.

**Comment by mykolapolonskyi** [comment]
@harshavardhana Maybe this issue can be converted to the feature to implement buckets creation from docker/docker-compose on start.
Seems to be pretty useful

**Comment by pun4drunk** [comment]
mounting a volume with subfolders automatically creates bucket(s) for you, works like a charm
```
volumes:
      - "./somefolder:/export"
```
Thank you @KanybekMomukeyev

**Comment by vladkras** [comment]
Why do you use `/export` folder? [Minio Docs](https://docs.minio.io/docs/minio-docker-quickstart-guide) says `/data`. Kinda trick?

**Comment by markhowells** [comment]
@vladkras You can use any directory name you like as long as there a volume mounted at that mount.

**Comment by flik** [comment]
@KanybekMomukeyev 
browser working fine with: http://localhost:9004/minio/login
docker-compose:
```
s3minio:
    image: minio/minio
    ports:
      - "9004:9000"
    volumes:
      - "./somefolder:/export"
    environment:
      - "MINIO_ACCESS_KEY=AKIAIOSFODNN7EXAMPLE"
      - "MINIO_SECRET_KEY=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY"
    command: server /export
```

**errors**:

Fatal  error: Uncaught exception 'Aws\S3\Exception\S3Exception' with message  'Error executing "CreateBucket" on "http://localhost:9004/testbucket";  AWS HTTP error: cURL error 7: Failed to connect to localhost port 9004:  Connection refused (see  http://curl.haxx.se/libcurl/c/libcurl-errors.html)'  GuzzleHttp\Exception\ConnectException: cURL error 7: Failed to connect  to localhost port 9004: Connection refused (see  http://curl.haxx.se/libcurl/c/libcurl-errors.html) in /var/www/html/public/v2/upload/vendor/aws/aws-sdk-php/src/WrappedHttpHandler.php on line 191
--

**Comment by flik** [comment]
```
<?php
date_default_timezone_set('America/Los_Angeles');
require 'vendor/autoload.php';

$s3 = new Aws\S3\S3Client([
    'version' => 'latest',
    'region'  => 'us-east-1',
    'endpoint' => 'http://localhost:9004',
    'use_path_style_endpoint' => true,
    'credentials' => [
            'key'    => 'AKIAIOSFODNN7EXAMPLE',
            'secret' => 'wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY',
        ],
]);

$bucket = 'testbucket';

// Create a bucket
$result = $s3->createBucket([
    'Bucket' => $bucket,
]);

```

**Comment by nitisht** [comment]
@flik can you try 'endpoint' => 'http://localhost:9004',

**Comment by vladkras** [comment]
@KanybekMomukeyev, are you sure localhost of your host machine is accessible from your php container? Usually, container's localhost is it's own 127.0.0.1 loop.

If both your containers are in the same docker-compse.yml file try `http://s3minio:9004`. If not, add with `docker network connect <s3minio_network> <php_container>` or vice versa

**Comment by KanybekMomukeyev** [comment]
@vladkras i think your question was related to @flik 🙂

**Comment by flik** [comment]
@vladkras  
I have done but same error:

Fatal error: Uncaught exception 'Aws\S3\Exception\S3Exception' with message 'Error executing "CreateBucket" on "http://s3minio:9004/testbucket"; AWS HTTP error: cURL error 7: Failed to connect to s3minio port 9004: Connection refused (see http://curl.haxx.se/libcurl/c/libcurl-errors.html)' GuzzleHttp\Exception\ConnectException: cURL error 7: Failed to connect to s3minio port 9004: Connection refused (see http://curl.haxx.se/libcurl/c/libcurl-errors.html) in /var/www/html/public/v2/upload/vendor/aws/aws-sdk-php/src/WrappedHttpHandler.php on line 191

**Comment by flik** [comment]
@KanybekMomukeyev  @vladkras May I add this script too ?

```
createbucket:
  image: garland/aws-cli-docker
  links:
    - minio
  command: >
    /bin/sh -c "
      while ! wget -q -s http://minio:9000/minio/login; do sleep 1; done;
      aws configure set default.s3.signature_version s3v4;
      aws configure set default.region us-east-1;
      aws s3api create-bucket \
        --bucket somebucketname \
        --endpoint-url=http://minio:9000;
      exit 0;
    "
  environment:
    - "AWS_ACCESS_KEY_ID=AKIAIOSFODNN7EXAMPLE"
    - "AWS_SECRET_ACCESS_KEY=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY"
```

**Comment by vladkras** [comment]
@KanybekMomukeyev yep, sorry, bro

@flik are you sure both containers are in same network (otherwise they cannot resolve each other by name)? Check it with `docker inspect <container_name>`

Also you can try to specify it's exact ip for test http://172.x.y.z:9004 which you can find out with `docker inspect ...` command too

**Comment by flik** [comment]
here detail:

```
docker inspect s3minio
[
    {
        "Id": "0739634a032e9f88c64f64a82c7b8e93b57e2f0f8d893b3ef8904507dbd6b544",
        "Created": "2018-11-15T12:18:02.8052182Z",
        "Path": "/usr/bin/docker-entrypoint.sh",
        "Args": [
            "server",
            "/export"
        ],
        "State": {
            "Status": "running",
            "Running": true,
            "Paused": false,
            "Restarting": false,
            "OOMKilled": false,
            "Dead": false,
            "Pid": 54014,
            "ExitCode": 0,
            "Error": "",
            "StartedAt": "2018-11-15T12:18:05.8287011Z",
            "FinishedAt": "0001-01-01T00:00:00Z",
            "Health": {
                "Status": "healthy",
                "FailingStreak": 0,
                "Log": [
                    {
                        "Start": "2018-11-15T12:21:06.2125436Z",
                        "End": "2018-11-15T12:21:06.2969437Z",
                        "ExitCode": 0,
                        "Output": "+ _init\n+ scheme=http://\n+ netstat -nplt\n+ awk  /(.*\\/minio)/ { gsub(\":::\",\"127.0.0.1:\",$4); print $4}\n+ address=0.0.0.0:9000\n+ resource=/minio/health/live\n+ stat -c %Y /proc/1\n+ start=1542284285\n+ healthcheck_main\n+ date +%s\n+ [ 181 -lt 120 ]\n+ curl -s -k -o /dev/null -w %{http_code} http://0.0.0.0:9000/minio/health/live\n+ http_response=200\n+ curl -k -s http://0.0.0.0:9000/minio/health/live\n+ http_response_body=\n+ [ 200 = 403 ]\n+ [ 200 = 200 ]\n"
                    },
                    {
                        "Start": "2018-11-15T12:21:36.2690876Z",
                        "End": "2018-11-15T12:21:36.3532003Z",
                        "ExitCode": 0,
                        "Output": "+ _init\n+ scheme=http://\n+ netstat -nplt\n+ awk  /(.*\\/minio)/ { gsub(\":::\",\"127.0.0.1:\",$4); print $4}\n+ address=0.0.0.0:9000\n+ resource=/minio/health/live\n+ stat -c %Y /proc/1\n+ start=1542284285\n+ healthcheck_main\n+ date +%s\n+ [ 211 -lt 120 ]\n+ curl -s -k -o /dev/null -w %{http_code} http://0.0.0.0:9000/minio/health/live\n+ http_response=200\n+ curl -k -s http://0.0.0.0:9000/minio/health/live\n+ http_response_body=\n+ [ 200 = 403 ]\n+ [ 200 = 200 ]\n"
                    },
                    {
                        "Start": "2018-11-15T12:22:06.323423Z",
                        "End": "2018-11-15T12:22:06.4076446Z",
                        "ExitCode": 0,
                        "Output": "+ _init\n+ scheme=http://\n+ netstat -nplt\n+ awk  /(.*\\/minio)/ { gsub(\":::\",\"127.0.0.1:\",$4); print $4}\n+ address=0.0.0.0:9000\n+ resource=/minio/health/live\n+ stat -c %Y /proc/1\n+ start=1542284285\n+ healthcheck_main\n+ date +%s\n+ [ 241 -lt 120 ]\n+ curl -s -k -o /dev/null -w %{http_code} http://0.0.0.0:9000/minio/health/live\n+ http_response=200\n+ curl -k -s http://0.0.0.0:9000/minio/health/live\n+ http_response_body=\n+ [ 200 = 403 ]\n+ [ 200 = 200 ]\n"
                    },
                    {
                        "Start": "2018-11-15T12:22:36.379209Z",
                        "End": "2018-11-15T12:22:36.4625302Z",
                        "ExitCode": 0,
                        "Output": "+ _init\n+ scheme=http://\n+ netstat -nplt\n+ awk  /(.*\\/minio)/ { gsub(\":::\",\"127.0.0.1:\",$4); print $4}\n+ address=0.0.0.0:9000\n+ resource=/minio/health/live\n+ stat -c %Y /proc/1\n+ start=1542284285\n+ healthcheck_main\n+ date +%s\n+ [ 271 -lt 120 ]\n+ curl -s -k -o /dev/null -w %{http_code} http://0.0.0.0:9000/minio/health/live\n+ http_response=200\n+ curl -k -s http://0.0.0.0:9000/minio/health/live\n+ http_response_body=\n+ [ 200 = 403 ]\n+ [ 200 = 200 ]\n"
                    },
                    {
                        "Start": "2018-11-15T12:23:06.4344435Z",
                        "End": "2018-11-15T12:23:06.5187943Z",
                        "ExitCode": 0,
                        "Output": "+ _init\n+ scheme=http://\n+ netstat -nplt\n+ awk  /(.*\\/minio)/ { gsub(\":::\",\"127.0.0.1:\",$4); print $4}\n+ address=0.0.0.0:9000\n+ resource=/minio/health/live\n+ stat -c %Y /proc/1\n+ start=1542284285\n+ healthcheck_main\n+ date +%s\n+ [ 301 -lt 120 ]\n+ curl -s -k -o /dev/null -w %{http_code} http://0.0.0.0:9000/minio/health/live\n+ http_response=200\n+ curl -k -s http://0.0.0.0:9000/minio/health/live\n+ http_response_body=\n+ [ 200 = 403 ]\n+ [ 200 = 200 ]\n"
                    }
                ]
            }
        },
        "Image": "sha256:6d641a0023d86b9f02b7feea86a0d7af095bb613d5e3d3e37a1511ca1e1930a2",
        "ResolvConfPath": "/var/lib/docker/containers/0739634a032e9f88c64f64a82c7b8e93b57e2f0f8d893b3ef8904507dbd6b544/resolv.conf",
        "HostnamePath": "/var/lib/docker/containers/0739634a032e9f88c64f64a82c7b8e93b57e2f0f8d893b3ef8904507dbd6b544/hostname",
        "HostsPath": "/var/lib/docker/containers/0739634a032e9f88c64f64a82c7b8e93b57e2f0f8d893b3ef8904507dbd6b544/hosts",
        "LogPath": "/var/lib/docker/containers/0739634a032e9f88c64f64a82c7b8e93b57e2f0f8d893b3ef8904507dbd6b544/0739634a032e9f88c64f64a82c7b8e93b57e2f0f8d893b3ef8904507dbd6b544-json.log",
        "Name": "/s3minio",
        "RestartCount": 0,
        "Driver": "overlay2",
        "Platform": "linux",
        "MountLabel": "",
        "ProcessLabel": "",
        "AppArmorProfile": "",
        "ExecIDs": null,
        "HostConfig": {
            "Binds": [
                "/Users/xdev/minio1-data:/export:rw"
            ],
            "ContainerIDFile": "",
            "LogConfig": {
                "Type": "json-file",
                "Config": {}
            },
            "NetworkMode": "xdev_webnet",
            "PortBindings": {
                "9000/tcp": [
                    {
                        "HostIp": "",
                        "HostPort": "9004"
                    }
                ]
            },
            "RestartPolicy": {
                "Name": "",
                "MaximumRetryCount": 0
            },
            "AutoRemove": false,
            "VolumeDriver": "",
            "VolumesFrom": [],
            "CapAdd": null,
            "CapDrop": null,
            "Dns": null,
            "DnsOptions": null,
            "DnsSearch": null,
            "ExtraHosts": null,
            "GroupAdd": null,
            "IpcMode": "shareable",
            "Cgroup": "",
            "Links": null,
            "OomScoreAdj": 0,
            "PidMode": "",
            "Privileged": false,
            "PublishAllPorts": false,
            "ReadonlyRootfs": false,
            "SecurityOpt": null,
            "UTSMode": "",
            "UsernsMode": "",
            "ShmSize": 67108864,
            "Runtime": "runc",
            "ConsoleSize": [
                0,
                0
            ],
            "Isolation": "",
            "CpuShares": 0,
            "Memory": 0,
            "NanoCpus": 0,
            "CgroupParent": "",
            "BlkioWeight": 0,
            "BlkioWeightDevice": null,
            "BlkioDeviceReadBps": null,
            "BlkioDeviceWriteBps": null,
            "BlkioDeviceReadIOps": null,
            "BlkioDeviceWriteIOps": null,
            "CpuPeriod": 0,
            "CpuQuota": 0,
            "CpuRealtimePeriod": 0,
            "CpuRealtimeRuntime": 0,
            "CpusetCpus": "",
            "CpusetMems": "",
            "Devices": null,
            "DeviceCgroupRules": null,
            "DiskQuota": 0,
            "KernelMemory": 0,
            "MemoryReservation": 0,
            "MemorySwap": 0,
            "MemorySwappiness": null,
            "OomKillDisable": false,
            "PidsLimit": 0,
            "Ulimits": null,
            "CpuCount": 0,
            "CpuPercent": 0,
            "IOMaximumIOps": 0,
            "IOMaximumBandwidth": 0,
            "MaskedPaths": [
                "/proc/acpi",
                "/proc/kcore",
                "/proc/keys",
                "/proc/latency_stats",
                "/proc/timer_list",
                "/proc/timer_stats",
                "/proc/sched_debug",
                "/proc/scsi",
                "/sys/firmware"
            ],
            "ReadonlyPaths": [
                "/proc/asound",
                "/proc/bus",
                "/proc/fs",
                "/proc/irq",
                "/proc/sys",
                "/proc/sysrq-trigger"
            ]
        },
        "GraphDriver": {
            "Data": {
                "LowerDir": "/var/lib/docker/overlay2/5232e390fc63f47461adf787c7357488d179f3ee376cdca63e584b4c47474842-init/diff:/var/lib/docker/overlay2/aef9046f27fc87faaf2c1c4b8826c0e4a5362ca6e3f89802c74d926bb9a36fe8/diff:/var/lib/docker/overlay2/4626a4ff245a597985e07f117b00ff36cade8a1f7449954b639730e45238714a/diff:/var/lib/docker/overlay2/9893641a617500bcf742a47392b1ba642566c228343e77d3e894795b24671f08/diff",
                "MergedDir": "/var/lib/docker/overlay2/5232e390fc63f47461adf787c7357488d179f3ee376cdca63e584b4c47474842/merged",
                "UpperDir": "/var/lib/docker/overlay2/5232e390fc63f47461adf787c7357488d179f3ee376cdca63e584b4c47474842/diff",
                "WorkDir": "/var/lib/docker/overlay2/5232e390fc63f47461adf787c7357488d179f3ee376cdca63e584b4c47474842/work"
            },
            "Name": "overlay2"
        },
        "Mounts": [
            {
                "Type": "bind",
                "Source": "/Users/xdev/minio1-data",
                "Destination": "/export",
                "Mode": "rw",
                "RW": true,
                "Propagation": "rprivate"
            },
            {
                "Type": "volume",
                "Name": "54e4a9e70cf2ac713d7aa95f7961116e14f7bb2370a800167938abc6ed0e73f5",
                "Source": "/var/lib/docker/volumes/54e4a9e70cf2ac713d7aa95f7961116e14f7bb2370a800167938abc6ed0e73f5/_data",
                "Destination": "/data",
                "Driver": "local",
                "Mode": "",
                "RW": true,
                "Propagation": ""
            }
        ],
        "Config": {
            "Hostname": "0739634a032e",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "ExposedPorts": {
                "9000/tcp": {}
            },
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": [
                "MINIO_ACCESS_KEY=AKIAIOSFODNN7EXAMPLE",
                "MINIO_SECRET_KEY=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY",
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
                "MINIO_UPDATE=off",
                "MINIO_ACCESS_KEY_FILE=access_key",
                "MINIO_SECRET_KEY_FILE=secret_key"
            ],
            "Cmd": [
                "server",
                "/export"
            ],
            "Healthcheck": {
                "Test": [
                    "CMD-SHELL",
                    "/usr/bin/healthcheck.sh"
                ],
                "Interval": 30000000000,
                "Timeout": 5000000000
            },
            "Image": "minio/minio",
            "Volumes": {
                "/data": {},
                "/export": {}
            },
            "WorkingDir": "",
            "Entrypoint": [
                "/usr/bin/docker-entrypoint.sh"
            ],
            "OnBuild": null,
            "Labels": {
                "com.docker.compose.config-hash": "fb54b3075154b321ace817b1a2b1ff0765e2f234b6d7a125e12859687f78479d",
                "com.docker.compose.container-number": "1",
                "com.docker.compose.oneoff": "False",
                "com.docker.compose.project": "xdev",
                "com.docker.compose.service": "s3minio",
                "com.docker.compose.version": "1.22.0",
                "maintainer": "Minio Inc <dev@minio.io>"
            }
        },
        "NetworkSettings": {
            "Bridge": "",
            "SandboxID": "f0c07192f58876a0b68976cc32da4f422c9b428a0b34273e9a40e5c09925a136",
            "HairpinMode": false,
            "LinkLocalIPv6Address": "",
            "LinkLocalIPv6PrefixLen": 0,
            "Ports": {
                "9000/tcp": [
                    {
                        "HostIp": "0.0.0.0",
                        "HostPort": "9004"
                    }
                ]
            },
            "SandboxKey": "/var/run/docker/netns/f0c07192f588",
            "SecondaryIPAddresses": null,
            "SecondaryIPv6Addresses": null,
            "EndpointID": "",
            "Gateway": "",
            "GlobalIPv6Address": "",
            "GlobalIPv6PrefixLen": 0,
            "IPAddress": "",
            "IPPrefixLen": 0,
            "IPv6Gateway": "",
            "MacAddress": "",
            "Networks": {
                "xdev_webnet": {
                    "IPAMConfig": {
                        "IPv4Address": "10.0.3.4"
                    },
                    "Links": null,
                    "Aliases": [
                        "s3minio",
                        "0739634a032e"
                    ],
                    "NetworkID": "t1q90r7muvy9u2vmoff2a8te1",
                    "EndpointID": "25d4bfc0164c2b45f4f45a30021136e7f241e689b992c82730dea6a9f5915365",
                    "Gateway": "",
                    "IPAddress": "10.0.3.4",
                    "IPPrefixLen": 24,
                    "IPv6Gateway": "",
                    "GlobalIPv6Address": "",
                    "GlobalIPv6PrefixLen": 0,
                    "MacAddress": "02:42:0a:00:03:04",
                    "DriverOpts": null
                }
            }
        }
    }
]
```

**Comment by flik** [comment]
"IPAddress": "10.0.3.4" Network ip is working. Issue **fixed**

**Comment by vladkras** [comment]
@flik be carefull, this ip will change each time your container is restarted. You should use some [networking](https://docs.docker.com/network/) to avoid such problems

**Comment by GvandeSteeg** [comment]
I know I'm several months out of date, but I figured I'd respons to an issue before raising a new one.
I've got a docker-compose file, but it doesn't set up my buckets for me.

```
minio_test_server:
    image: minio/minio:RELEASE.2019-03-27T22-35-21Z
    volumes:
      - ./haem/test_haem:/data
    ports:
      - "9000:9000"
    environment:
      MINIO_ACCESS_KEY: haem_admin
      MINIO_SECRET_KEY: haem_admin
    command: gateway nas /data
```

Where `./haem/test_haem/` contains a folder `logfiles`, which contains a file `passing_example.trc`
I use PyCharm to run my docker compose.
When I run this, my Minio browser is completely empty, but running the equivalent through Docker works like a charm
